internal/storage/fs/snapshot.go

```diff
@@ -1,14 +1,616 @@
+			}
+		}
+	}
+
+	return filenames, nil
+}
+
+func (ss *storeSnapshot) addDoc(doc *ext.Document) error {
+	ns := ss.ns[doc.Namespace]
+	if ns == nil {
+		ns = newNamespace(doc.Namespace)
+	}
+
+	evalDists := map[string][]*storage.EvaluationDistribution{}
+
+	now := timestamppb.Now()
+
+	for _, s := range doc.Segments {
+		matchType := flipt.MatchType_value[s.MatchType]
+		segment := &flipt.Segment{
+			NamespaceKey: doc.Namespace,
+			Name:         s.Name,
+			Key:          s.Key,
+			Description:  s.Description,
+			MatchType:    flipt.MatchType(matchType),
+			CreatedAt:    now,
+			UpdatedAt:    now,
+		}
+
+		for _, constraint := range s.Constraints {
+			constraintType := flipt.ComparisonType_value[constraint.Type]
+			segment.Constraints = append(segment.Constraints, &flipt.Constraint{
+				NamespaceKey: doc.Namespace,
+				SegmentKey:   segment.Key,
+				Id:           uuid.Must(uuid.NewV4()).String(),
+				Operator:     constraint.Operator,
+				Property:     constraint.Property,
+				Type:         flipt.ComparisonType(constraintType),
+				Value:        constraint.Value,
+				CreatedAt:    now,
+				UpdatedAt:    now,
+			})
+		}
+
+		ns.segments[segment.Key] = segment
+	}
+
+	for _, f := range doc.Flags {
+		flag := &flipt.Flag{
+			NamespaceKey: doc.Namespace,
+			Key:          f.Key,
+			Name:         f.Name,
+			Description:  f.Description,
+			Enabled:      f.Enabled,
+			CreatedAt:    now,
+			UpdatedAt:    now,
+		}
+
+		for _, v := range f.Variants {
+			attachment, err := json.Marshal(v.Attachment)
+			if err != nil {
+				return err
+			}
+
+			flag.Variants = append(flag.Variants, &flipt.Variant{
+				Id:           uuid.Must(uuid.NewV4()).String(),
+				NamespaceKey: doc.Namespace,
+				Key:          v.Key,
+				Name:         v.Name,
+				Description:  v.Description,
+				Attachment:   string(attachment),
+				CreatedAt:    now,
+				UpdatedAt:    now,
+			})
+		}
+
+		ns.flags[f.Key] = flag
+
+		evalRules := []*storage.EvaluationRule{}
+		for _, r := range f.Rules {
+			rule := &flipt.Rule{
+				NamespaceKey: doc.Namespace,
+				Id:           uuid.Must(uuid.NewV4()).String(),
+				FlagKey:      f.Key,
+				SegmentKey:   r.SegmentKey,
+				Rank:         int32(r.Rank),
+				CreatedAt:    now,
+				UpdatedAt:    now,
+			}
+
+			evalRule := &storage.EvaluationRule{
+				NamespaceKey: doc.Namespace,
+				FlagKey:      f.Key,
+				ID:           rule.Id,
+				Rank:         rule.Rank,
+				SegmentKey:   rule.SegmentKey,
+			}
+
+			segment := ns.segments[rule.SegmentKey]
+			if segment == nil {
+				return ferrors.ErrNotFoundf("segment %q in rule %d", rule.SegmentKey, rule.Rank)
+			}
+
+			evalRule.SegmentMatchType = segment.MatchType
+
+			for _, constraint := range segment.Constraints {
+				evalRule.Constraints = append(evalRule.Constraints, storage.EvaluationConstraint{
+					Operator: constraint.Operator,
+					Property: constraint.Property,
+					Type:     constraint.Type,
+					Value:    constraint.Value,
+				})
```

Comment:
I originally implemented this before constraint description.
Could you add those back in here? (plus maybe a test case to be sure).